Render widget areas referencing blocks in a wp_area post on the website frontend

jorgefilipecosta opened the pull request:
## Description
Here we implement a simple filter that makes a sidebar referencing a post_id reference a simple widget that renders the content of the post.

This is the first version and the most straightforward approach. To improve back-compatibility, I think we can parse the post. And replace legacy widget blocks to point to the original widget they referred, this change will allow users that go the previous screen to see the widgets they had there as they would typically see.

## How has this been tested?
I executed this code in the browser console:
```
wp.apiFetch({
    path: '/__experimental/widget-areas/sidebar-1',
    data: { content: '<!-- wp:columns --><div class="wp-block-columns has-2-columns"><!-- wp:column --><div class="wp-block-column"><!-- wp:paragraph --><p>Col 1</p><!-- /wp:paragraph --></div><!-- /wp:column --><!-- wp:column --><div class="wp-block-column"><!-- wp:paragraph --><p>Col 2</p><!-- /wp:paragraph --></div><!-- /wp:column --></div><!-- /wp:columns -->' },
    method: 'POST',
}).then(console.log);
```
I went to the website frontend and verified the blocks appeared there.

jorgefilipecosta committed:
revisions (+2 squashed commits)
Squashed commits:
[fa58e1f32] Update lib/class-experimental-wp-widget-blocks-manager.php

Co-Authored-By: Robert Anderson <robert@noisysocks.com>
[ee18b08f2] Render widget areas referencing blocks in a wp_area post on the website frontend.

noisysocks commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -245,6 +257,9 @@ public static function serialize_blocks( $blocks ) {
 	 * @return string String representing the block.
 	 */
 	public static function serialize_block( $block ) {
+		if( ! isset( $block[ 'blockName' ] ) ) {
```
Spacing is not quite right here. 

```suggestion
		if ( ! isset( $block['blockName'] ) ) {
```

noisysocks commented on lib/widgets.php:
```diff
@@ -181,3 +181,88 @@ function gutenberg_create_wp_area_post_type() {
+ * @param array $blocks   Array of blocks.
+ */
+function gutenberg_blocks_to_widget( $blocks ) {
+	$widget_id = 'blocks-widget-' . md5( Experimental_WP_Widget_Blocks_Manager::serialize_block( $blocks ) );
```
Let's pull this function out of `Experimental_WP_Widget_Blocks_Manager`. We already have `serialize_blocks` so it makes sense in my mind that we'd have `serialize_block` as well.

jorgefilipecosta replied:
Hi @noisysocks, in other PR @youknowriad recommended that we keep serialize_blocks inside Experimental_WP_Widget_Blocks_Manager to reinforce the experimental nature of the function. Given that recomendation, I guess we should also keep function gutenberg_blocks_to_widget( $blocks ) inside.

youknowriad replied:
Also what does serializing means? Is it's the same as the frontend? If it is, it means the blocks should be fully parsed and not just half-parsed.

jorgefilipecosta replied:
I think the serialization we have here is very simple and may not even work on all cases yet. It only needs to work for legacy widgets. I guess it is better to keep it inside the experimental class. And then we may have to rethink this to simply have functions like serialize_legacy_widgets which make their objectives more clear and may end up being public.

noisysocks commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -245,6 +257,9 @@ public static function serialize_blocks( $blocks ) {
 	 */
 	public static function serialize_block( $block ) {
+		if( ! isset( $block[ 'blockName' ] ) ) {
+			return '';
```
Usually in PHP one returns `false` when there's an error. It should not impact `serialize_blocks` because PHP automatically converts `false` to `''` when casting to a string.

```
php > var_dump( 'hello' . false . 'jorge' );
string(10) "hellojorge"
```

noisysocks reviewed (changes requested):
Nice work! Most of this looks familiar from my `try/` branch, so of course I'm happy 🙂

When I test I see a PHP warning:

<img width="463" alt="Screen Shot 2019-05-23 at 15 44 51" src="https://user-images.githubusercontent.com/612155/58228364-cddb7980-7d71-11e9-8cf0-6aecebc5abbf.png">

It looks like `wp_widget_control()` in `wp-admin/includes/widget.php` accesses `global $sidebars_widgets;` directly instead of calling `wp_get_sidebars_widgets()`. 

Can we work around this? Ideally `global $sidebars_widgets` should not contain the new data format so that it remains backwards compatible. Perhaps we could do a `$sidebars_widgets = gutenberg_swap_out_sidebars_blocks_for_block_widgets( $sidebars_widgets );` somewhere—possibly in the `init` action.

jorgefilipecosta commented:
> It looks like wp_widget_control() in wp-admin/includes/widget.php accesses global $sidebars_widgets; directly instead of calling wp_get_sidebars_widgets().

> Can we work around this? Ideally global $sidebars_widgets should not contain the new data format so that it remains backwards compatible. Perhaps we could do a $sidebars_widgets = gutenberg_swap_out_sidebars_blocks_for_block_widgets( $sidebars_widgets ); somewhere—possibly in the init action.

I was not being able to reproduce the problem may be a  plugin/theme is changing some behavior?
I tried to set the global in the init action but that's irrelevant as wp_get_sidebars_widgets changes the global variable anyway. I updated the filter to make sure the variable is also updated by the filter so if someone uses the global directly it should see the old structure. Let me know if this problem is persisting.


I also tried to use a lower level version where we filter directly using the filter option_sidebars_widgets so that we would change the result of get_option( 'sidebars_widgets' ) directly. That approach proved to be unfruitful given the mechanism of the options where we have cache the save become intermittent, for example saving in our endpoint sometimes caused the blocks to save as a widget block.

noisysocks commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -271,4 +292,93 @@ public static function serialize_block( $block ) {
+	 * @param array $options   Widget options.
+	 * @param array $arguments Arguments array.
+	 */
+	public static function gutenberg_output_blocks_widget( $options, $arguments ) {
```
Now that these methods are inside the experimental class, they don't need the `gutenberg_` prefix.

noisysocks replied:
Changed in 300e1c250.

noisysocks replied:
Renamed in 300e1c250.

noisysocks commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -271,4 +292,93 @@ public static function serialize_block( $block ) {
+	 * @param array $blocks   Array of blocks.
+	 */
+	public static function gutenberg_blocks_to_widget( $blocks ) {
+		$widget_id = 'blocks-widget-' . md5( Experimental_WP_Widget_Blocks_Manager::serialize_block( $blocks ) );
```
Could be simply `self::serialize_block`.

noisysocks replied:
Done in 2431bf6c2.

noisysocks commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -271,4 +292,93 @@ public static function serialize_block( $block ) {
+	 */
+	public static function gutenberg_swap_out_sidebars_blocks_for_block_widgets( $sidebars_widgets_input ) {
+		global $sidebars_widgets;
+		if ( null === self::$unfiltered_sidebar_widgets ) {
```
If `gutenberg_swap_out_sidebars_blocks_for_block_widgets` is called again with a different `$sidebars_widgets_input` then this won't update its cached copy of `self::$unfiltered_sidebar_widgets`. 

We probably want:

```php
if ( $sidebars_widgets_input !== self::$unfiltered_sidebar_widgets ) {
```

noisysocks replied:
Actually, not sure about this. Maybe it's not a problem. Is `global $sidebars_widgets` ever updated?

noisysocks reviewed (commented):
Yet again, I'm now not getting that error! Something must have been weird in my dev environment yesterday. 

This looks good to get in and iterate on. There's a few minor suggestions that I'll fix up myself.

noisysocks committed:
Remove unnecessary gutenberg_ prefix

noisysocks committed:
Simplify call to self::serialize_blocks

youknowriad commented:
Awesome work on all these widget screen related PRs @noisysocks @jorgefilipecosta
